### ISSUE TITLE
gcp.storage.BucketObject does not refresh an object if using the source parameter instead of content

### ISSUE THREAD

**Issue by sylv3r** [pulumi/pulumi-gcp#87]
Discussion on this matter with @clstokes on slack here : https://pulumi-community.slack.com/archives/C84L4E3N1/p1548693619818500

### Problem

When creating an object with the `source` parameter on a `gcp.storage.BucketObject` pulumi (or the provider) does not detect a change on the local file and thus does not update the related bucket object.

### How to reproduce

1. Create an object in a gcp bucket referencing a local file : 
```
new gcp.storage.BucketObject('my-object', {
  bucket: 'bucket-name',
  name: 'my-object',
  source: 'absolute/path/to/file.ext',
});
```
2. `pulumi up` succeeds to create the object in the bucket
3. Update the local file content (name and path stay unchanged, only the md5 is different)
4. Run `pulumi up` again -> no change detected

**Comment by joeduffy** [comment]
This definitely seems like a bug where we're accepting a `string` when only an `Archive` should be allowed. This would explain the odd diffing behavior.

**Comment by sylv3r** [comment]
@joeduffy Ok, I see what you mean when looking into the method documentation.
 
How one is supposed to create/use an `Archive` in this case instead of a `string` then ?

Couldn't you accept a `string` and make an automatically create an `Archive` (whatever that is) from it under the hood to handle this simple case ?

**Comment by clstokes** [comment]
Per an internal conversation, `source` should be retyped to `Asset` rather than `Archive`. This should achieve the correct/expected behavior.

**Comment by lukehoban** [comment]
See also https://github.com/pulumi/pulumi/issues/280 which is sort of a root cause here.

**Comment by lukehoban** [comment]
We should do a sweep of `source` properties that map to files on disk and ensure we have mapped them to `Assets`/`Archives`.  It was raised recently that `aws.lambda.Layer` has a similar problem (we addressed this for `aws.lambda.Function` a long time ago).  We'll also need to come up with a general strategy here on two fronts:
1. How to make the change from `source: string` to `something: Asset`
2. How to decide whether to accept `Asset` or `Archive`, and ideally how to ensure that we don't have to make this decision at all by fixing https://github.com/pulumi/pulumi/issues/280.

Given how many commonly used APIs this impacts, would be great to tackle these sooner than later.
